llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8789,8 +9029,535 @@ static bool combineCCMask(SDValue &CCReg, int &CCValid, int &CCMask) {
+    int Op0CCValidVal = Op0CCValid->getZExtValue();
+    if (!Op0TrueVal || !Op0FalseVal || !Op0CCMask || !Op0CCValid)
+      return SDValue();
+
+    const auto isOneBitSet = [](int Mask) {
+      return Mask && !(Mask & (Mask - 1));
+    };
+
+    // Only one-bit set.
+    const auto log2CCMaskToCCVal = [](int Mask) {
+      size_t Pos = 0;
+      while (!(Mask & 1)) {
+        Mask >>= 1;
+        ++Pos;
+      };
+      assert(Pos < 4 && "CC out of range");
+      return (3 - Pos);
+    };
+
+    int CCMask = Op0CCMask->getZExtValue();
+    SDValue TrueVal, FalseVal;
+    // There are two cases - either only one bit set or one clear bit.
+    if (isOneBitSet(CCMask)) {
+      int CCVal = log2CCMaskToCCVal(CCMask);
+      // This transformations is valid if Op0FalseVal is false for now.
+      // It can be extended for other Op0FalseVal value by replacing
+      // FalseVal with (XOR (N->getOperand(1))(Op0FalseVal)).
+      if (Op0FalseVal->getZExtValue() != 0)
+        return SDValue();
+      // select_ccmask (CCVal ^ Op0TrueVal) (SRL IPM CC) CCValid/CCMask CC.
+      TrueVal = DAG.getTargetConstant((CCVal ^ Op0TrueVal->getZExtValue()),
+                                      SDLoc(N), MVT::i32);
+      FalseVal = N->getOperand(1);
+    } else if (isOneBitSet((CCMask ^ 0xf))) {
+      // Only one clear bit.
+      // select_ccmask (XOR (SRL IPM CC) (Op0TrueVal)) (CCVal ^ Op0FalseVal)
+      // CCValid/CCMask CC.Value cc-value must have in FalseVal.
+      int CCVal = log2CCMaskToCCVal(CCMask ^ 0xf);
+      TrueVal = DAG.getNode(ISD::XOR, SDLoc(N), MVT::i32, N->getOperand(1),
+                            DAG.getTargetConstant(Op0TrueVal->getZExtValue(),
+                                                  SDLoc(N), MVT::i32));
+      FalseVal = DAG.getTargetConstant((CCVal ^ Op0FalseVal->getZExtValue()),
+                                       SDLoc(N), MVT::i32);
+    } else
+      return SDValue();
+
+    // Check Xor's user's user if it it select_ccmask or br_ccmask, put it
+    // back to WorkList so that algortirm can start processing from outer
+    // select_ccmask/br_ccmask and combine with (icmp (select_ccmask)).
+    for (SDUse &XorUse : N->uses()) {
+      auto *CCNode = XorUse.getUser();
+      if (CCNode && (CCNode->getOpcode() == SystemZISD::TM /*||
+                     CCNode->getOpcode() == SystemZISD::ICMP*/)) {
+        for (SDUse &SelectBrUse : CCNode->uses()) {
+          auto *SelectBr = SelectBrUse.getUser();
+          if (SelectBr && (SelectBr->getOpcode() == SystemZISD::SELECT_CCMASK ||
+                           SelectBr->getOpcode() == SystemZISD::BR_CCMASK))
+            DCI.AddToWorklist(SelectBr);
+        }
+      }
+    }
+    return DAG.getNode(
+        SystemZISD::SELECT_CCMASK, SDLoc(N), N->getValueType(0), TrueVal,
+        FalseVal, DAG.getTargetConstant(Op0CCValidVal, SDLoc(N), MVT::i32),
+        DAG.getTargetConstant(CCMask, SDLoc(N), MVT::i32), CCReg);
+  }
+  return SDValue();
+}
```

Comment:
In general, those three functions are now getting quite complicated, and at the same time more and more similar to each other.  I think it would be great if we could make them fully identical (except for uses of AND/OR/XOR) and then merge into a single function.